casper/src/main/scala/io/casperlabs/casper/util/execengine/ExecEngineUtil.scala

```
@@ -75,37 +81,28 @@ object ExecEngineUtil {
-        (deploy, cost, Some(eff))
-    }
-
   //TODO: Logic for picking the commuting group? Prioritize highest revenue? Try to include as many deploys as possible?
   def findCommutingEffects(
-      deployEffects: Seq[(Deploy, Long, Option[ExecutionEffect])]
-  ): Seq[(Deploy, Long, Option[ExecutionEffect])] = {
-    val (errors, errorFree) = deployEffects.span(_._3.isEmpty)
-
-    val nonConflicting = errorFree.toList match {
-      case Nil                           => Nil
-      case (head @ (_, _, eff0)) :: tail =>
-        // the `eff0.get` call is safe because of the `span` above which separates into Some and None cases
-        val (result, _) = tail.foldLeft(Vector(head) -> Op.fromIpcEntry(eff0.get.opMap)) {
-          case (unchanged @ (acc, totalOps), next @ (_, _, Some(eff))) =>
-            val ops = Op.fromIpcEntry(eff.opMap)
-            if (totalOps ~ ops)
-              (acc :+ next, totalOps + ops)
-            else
-              unchanged
-        }
+      deployEffects: Seq[ProcessedDeployResult]
+  ): Seq[ProcessedDeployResult] = {
+    // All the deploys that do not change the global state in a way that can conflict with others:
+    // + `ExecutionError` deploys update the nonce,
+    // + `PreconditionFailure` deploys have no effects
+    val (effectFree, effectful) = deployEffects.partition(!_.isInstanceOf[ExecutionSuccessful])
+
+    val nonConflicting = effectful match {
+      case Nil => List.empty[ProcessedDeployResult]
+      case list =>
+        val (result, _) =
+          list.foldLeft(List.empty[ProcessedDeployResult] -> Map.empty[ipc.Key, Op]) {
+            case (unchanged @ (acc, totalOps), next @ ExecutionSuccessful(_, effects, _)) =>
+              val ops = Op.fromIpcEntry(effects.opMap)
+              if (totalOps ~ ops)
+                (next :: acc, totalOps + ops)
+              else
+                unchanged
+            case (unchanged, _) => unchanged
```

The `partition` above already filtered for `ExecutionSuccessful`, can this ever match? Can you raise an error instead to signify that something went horribly wrong?

---

It can't but the encoding doesn't allow for type-level partition.

---

So why not do a `case _ => sys.error("Will never happen.")` just in case somebody changes the upper bit but forgets to update this one, so we don't silently exclude something.

---

I will remove that `case` and it will fail fast during tests/demos and we can fix it.